### ISSUE TITLE
parallel-letter-frequency: steep learning curve

### ISSUE THREAD

**Issue by Eptarch** [exercism/go#2067]
First of all, thanks for Exercism, it's an awesome tool to learn new programming language.

I have just ended the torture and cheated my way out of the parallel-letter-frequency exercise. It was my 33rd exercise in Go track and I find it to be **the worst** I had so far in terms of understanding what to do.

I'd like to illustrate my point:
1. Check out example solution at https://github.com/exercism/go/issues/1941#issuecomment-963046147
There is no way I'd write such a code given one line of requirements in the exercise ("Count the frequency of letters in texts using parallel computation.")
There is absolutely no way I could write anything like it without proper experience with Go and its approach to concurrency (Wait groups, concurrency patterns, ranging over channels, anonymous functions - I knew nothing of that).
2. There are zero hints. There's wall of text that leads you to read here and there, but after reading and watching 4 hours worth of information on concurrency and having some kind of understanding about what it is all about, I still had no idea how to approach the problem. My initial solution used worker approach, but I had no idea that iterating over opened channel never ends, so I got stuck again.
3. The previous exercise "the-clock" has 4492 people submitted their solutions, whereas "parallel-letter-frequency" has 2058. I don't know if it tells anything, the one before that is "grains" and it has 2786. Still, it looks like this exercise peels off some people off of the track.

There's quite a bit of concepts to utilize in this exercise, and it may be better to split it into smaller assignments.

PS I'm sorry, there maybe a better way to deliver feedback, but the exercise got me too frustrated to look it up :(

**Comment by junedev** [comment]
@Eptarch Thanks for pointing this out. We launched v3 of Exercism a bit more than 3 month ago. v3 theoretically allows us to define which learning exercises need to be completed before a practice exercise will be unlocked. Currently, we are working on creating these learning exercises for the different concepts. The ones related to concurrency are not created yet, see our roadmap https://exercism.org/docs/tracks/go/roadmap#h-potential-future-concepts. Creating concept documentation and inventing new learning exercises is a lot of work and everyone does this on their free time so it takes quite long to progress here. Once we have the learning exercises that teach about concurrency in place, we will set `parallel-letter-frequency` to only unlock after those are completed. Then everyone doing the exercise should be well equipped to solve it.

In the meantime we will discuss whether we can raise the difficulty setting here so it comes up a bit later in the exercise list.
In general, please keep in mind that there is no "blocking" mechanism in v3 of Exercism anymore. You can leave aside any practice exercise you like and just do another one instead.

Opening an issue for something that doesn't work out well is fine but please keep the [CoC](https://exercism.org/docs/using/legal/code-of-conduct#h-the-simple-version) in mind which states we try to be emphatic, kind and polite to each to other and remember that almost all the work on Exercism is done by volunteers.

@exercism/go Currently this exercise has difficulty 3 (same as e.g. anagram, acronym, beer song) which marks it as "easy" on the website. What about we raise the difficulty score for now so it shows as "medium", at least until we got all the unlocking in place?

**Comment by andrerfcsantos** [comment]
@Eptarch Thanks for our feedback. It's always great to have the feedback of people going through the track and talk about their pain points. This allow us to make things better, that's why I appreciate the time you took to let us know about your experience.

> There is absolutely no way I could write anything like it without proper experience with Go and its approach to concurrency (Wait groups, concurrency patterns, ranging over channels, anonymous functions - I knew nothing of that).

This is on us. Right now, we don't have any concept exercises teaching about this stuff, which I understand that does require someone to learn about this outside Exercism. In the future, we want to have concept exercises that cover these topics. The idea is for a new student on the track to complete the concept exercises first and then go over the practice exercises (parallel-letter-frequency is a practice exercise). Do you think that if you knew about these concepts before you would be able to figure out what the exercise intends?

> My initial solution used worker approach, but I had no idea that iterating over opened channel never ends, so I got stuck again.

One of the things I love about this exercise is how different the solutions can be. In the most simple solution, you just have to know about channels and goroutines, everything else is optional to get a first solution working. In such solutions, you don't need to know about waitgroups, ranges over channels, anonymous functions or even that channels can be closed. A mentor can then help to introduce more concepts bit by bit into the solution.

> The previous exercise "the-clock" has 4492 people submitted their solutions, whereas "parallel-letter-frequency" has 2058. I don't know if it tells anything, the one before that is "grains" and it has 2786. Still, it looks like this exercise peels off some people off of the track.

That's an interesting stat.

> Currently this exercise has difficulty 3 (same as e.g. anagram, acronym, beer song) which marks it as "easy" on the website. What about we raise the difficulty score for now so it shows as "medium", at least until we got all the unlocking in place?

I agree. I don't think the exercise itself is hard, but we should raise the difficulty just based on the amount of concepts someone must be familiar with to solve it.

**Comment by lburnett88** [comment]
Hello,

I just completed this was some mentoring from my boyfriend, and agree it was harder to work out on my own. However the bigger problem was that my first solution wasn't actually concurrent, I had incorrectly used a range loop and set the functions off sequentially - but still passed all tests.

My boyfriend pointed this out, but looking at community solutions it looks like lots of people fell into the same trap.

Note - I am a programming beginner beyond Exercism so my understanding of the concepts is very fuzzy!

L

**Comment by junedev** [comment]
@lburnett88 Thanks for leaving a note here.
In theory, you are supposed to use the benchmarks to check whether your solution performs better then the sequential solution. That will tell you how good you leverage concurrent execution. However, currently the test cases are not very good for this. We have a separate issue about this here: https://github.com/exercism/go/issues/1941
Unfortunately, we cannot run the benchmarks on the website currently so you have to work locally to see the benchmark results. The usual test cases cannot distinguish a sequential from a concurrent solution, they only check whether the output is correct.

I think we should clarify how to validate the solution in the instructions for the this exercise. I will leave a note about that in the other issue. Thanks again for pointing out this problem.

Please keep in mind that Exercism is not the perfect platform for programming beginners. It is designed for people that already know one programming language well and want to learn another one. That is why you will probably encounter some content that is hard to understand. Unfortunately, I don't know a good resource to learn Go for programming beginners, otherwise I would give some recommendation.

**Comment by andrerfcsantos** [comment]
> I had incorrectly used a range loop and set the functions off sequentially - but still passed all tests.
> My boyfriend pointed this out, but looking at community solutions it looks like lots of people fell into the same trap.

@lburnett88  Can you share with us what such a sequential solution looks like or an example of a published solution that is sequential but passes the tests? I went through the first page of the most starred solutions (30 solutions) and all of them use goroutines. The same happens when searching for the newest solutions.

**Comment by lburnett88** [comment]
> 

@andrerfcsantos My first solution [here](https://exercism.org/tracks/go/exercises/parallel-letter-frequency/solutions/lburnett88) - it does use goroutines, but each goroutine ran one at a time as I'd used a channel to block the result, so although using goroutines it wasn't actually parallelised/concurrent. 

@junedev thanks. I tried to get exercism running locally, but struggled with the instructions as I have an apple macbook with an apple silicon chip, not an intel processor and it seemed the instructions didn't work well for that. It's on my 'to do' list to revisit :)

**Comment by andrerfcsantos** [comment]
@lburnett88 Ah, I see, thanks for sharing. However, I'm not sure if we can solve that particular mistake by adding something to the instructions. I feel like this exercise has a lot of little details that can really benefit from a mentor, and asking for a mentor review is probably the best thing to do for now while we don't have the exercises practicing these basic concepts separately.

Meanwhile, we increased the difficulty of the exercise to "medium", hopefully that works as a warning to people that the exercise might require some work.

Thanks again for the feedback.

**Comment by tehsphinx** [comment]
@lburnett88: Just a note on the solution: The goroutines are not running one at a time. Only the results are sent back one at a time. If the `Frequency` function would do a lot of work (like a few seconds of work, and there'd be not too many of them), that would distribute the work to all available CPUs just fine :)
